src/mongocxx/test/client_side_encryption.cpp

```diff
@@ -322,28 +360,569 @@ TEST_CASE("Datakey and double encryption", "[client_side_encryption]") {
+                                      << "unencrypted" << under_16mib << finalize;
+
+    // Expect this to succeed since this is still (just) under the maxBsonObjectSize limit.
+    coll.insert_one(under_16mib_doc.view());
+
+    // Insert the document limits/limits-doc.json concatenated with
+    // { "_id": "encryption_exceeds_16mib",
+    //   "unencrypted": < the string 'a' repeated (16777216 - 2000) times > }
+    auto encryption_exceeds_16mib = document{} << "_id"
+                                               << "encryption_exceeds_16mib"
+                                               << "unencrypted" << under_16mib
+                                               << concatenate(limits_doc.view()) << finalize;
+
+    // Expect this to fail since encryption results in a document exceeding
+    // the maxBsonObjectSize limit.
+    REQUIRE_THROWS(coll.insert_one(encryption_exceeds_16mib.view()));
 }
 
 TEST_CASE("Views are prohibited", "[client_side_encryption]") {
-    // TODO CXX-2021
+    instance::current();
+
+    // Create a MongoClient without encryption enabled (referred to as client).
+    class client client {
+        uri {}
+    };
+
+    if (test_util::get_max_wire_version(client) < 8) {
+        // Automatic encryption requires wire version 8.
+        WARN("Skipping - max wire version is < 8");
+        return;
+    }
+
+    // Using client, drop and create a view named db.view with an empty pipeline.
+    // E.g. using the command { "create": "view", "viewOn": "coll" }.
+    auto db = client["db"];
+    db["view"].drop();
+
+    auto cmd = make_document(kvp("create", "view"), kvp("viewOn", "coll"));
+    db.run_command(cmd.view());
+
+    // Create a MongoClient configured with auto encryption (referred to as client_encrypted)
+    // Configure with the local KMS provider as follows:
+    // { "local": { "key": <base64 decoding of LOCAL_MASTERKEY> } }
+    // Configure with the keyVaultNamespace set to keyvault.datakeys.
+    options::client opts;
+    auto kms_doc = _make_kms_doc();
+    _add_client_encrypted_opts(&opts, {}, std::move(kms_doc));
+    class client client_encrypted {
+        uri{}, std::move(opts)
+    };
+
+    // Using client_encrypted, attempt to insert a document into db.view.
+    // Expect an exception to be thrown containing the message: "cannot auto encrypt a view".
+    REQUIRE_THROWS_MATCHES(client_encrypted["db"]["view"].insert_one(cmd.view()),
+                           mongocxx::exception,
+                           test_util::mongocxx_exception_matcher{"cannot auto encrypt a view"});
 }
 
 TEST_CASE("Corpus", "[client_side_encryption]") {
-    // TODO CXX-2021
+    // TODO CXX-2022
+}
+
+void _round_trip(class client_encryption* client_encryption,
+                 bsoncxx::types::bson_value::view datakey) {
+    auto to_encrypt_doc = document{} << "v"
+                                     << "test" << finalize;
+    auto to_encrypt = to_encrypt_doc.view()["v"].get_value();
```

Comment:
So, this line
`auto to_encrypt = to_encrypt_doc.view()["v"].get_value();`

 is actually dealing with four data types: the owning `document::value` `to_encrypt_doc`, the temporary `document::view` `to_encrypt_doc.view()`, a temporary `document::element` that is created when we use `["v"]`, and the final captured `bson_value::view` `to_encrypt`.

I see no problem with this flow; even though the `document::view` and the `document::element` in the chain are temporary objects.  Both of those temporaries are views into the owning `to_encrypt_doc`'s buffer, so there's no lifetime issues when these temporaries go away.  In fact, I think this is/should be one of our "recommended" flows, since it utilizes the view-ness of views in order to avoid copies and moves.

That being said, I rather hate the construction `doc_value.view()["field"]`, and I'm excited to make it better as part of [CXX-2046](https://jira.mongodb.org/browse/CXX-2046).

This flow uses a similar "chain of temporaries" and is how we recommend users access bson fields:
```
auto doc_value = document{} << "key" << "value" << finalize; //1
auto doc_view = doc_value.view(); //2
auto string_val = doc_view["key"].get_utf8().value; //3
```
Line `//3` that pulls out the string is creating two temporary non-owning view types, the flow is `document::view` -> `document::element` -> `bsoncxx::types::b_utf8` -> `stdx::string_view`.  Here, yes, the `document::view doc_view` is not a temporary in this one-liner, as in the line of code in question, but it's still not owning any of the data.  Pulling it out into a named variable on line `//2` is really just a convenience that makes the code a little more readable, IMO.

I would very much like to discourage users from creating the b_type structs manually.  The b_type structs are inherently view-y and have no owning counterpart, so if users create their own manually they must manage the lifetime of all the component parts correctly.  To me this sounds like a nice foot-gun.  While, yes, this code that you've written will be correct:
```
std::string str = "test";
bsoncxx::bson_value::view to_encrypt {bsoncxx::types::b_utf8 {str}};
```
I think it's much more likely that users will write this and then have a problem:
`bsoncxx::bson_value::view to_encrypt {bsoncxx::types::b_utf8 {"test"}};`

I would rather encourage users to work with `document` as the starting point whenever possible, even though it's a little roundabout, since it will make the ownership semantics easier to reason about and to do correctly:
```
auto doc = make_document(kvp("v", "test")); // or, auto doc = document{} << "v" << "test" << finalize;
auto doc_view = doc.view(); // optional intermediate step
auto to_encrypt = doc_view["v"].get_value(); // this is a bson_value::view
```

If they want to construct an owning bson_value type, they have two options:
```
// 1: use bson_value::value ctor to convert to an owning bson_value
bson_value::value to_encrypt_owning{to_enrypt};
// 2: call get_owned_value() instead of get_value(), when accessing the bson_value in the first place
auto to_encrypt_owning = doc_view["v"].get_owning_value();
```

Reply:
I like that helper idea!  I created a ticket for it: [CXX-2066](https://jira.mongodb.org/browse/CXX-2066).  I agree, it is roundabout, and I think that's a shortcoming of our bson value/document type system hierarchy.  It would be great to fix this in the future, so owning `bson_value`s can be easily created without going through the document builder.

As an improvement after the release, I'd like to expose constructors for `bson_value::value` outright.  I see these constructors as taking the component parts of each `b_type` struct, but not actually using the `b_type` structs.  To me the `b_type` structs are problematic because they are inherently non-owning, and it requires a copy to make owning `bson_value`'s from them.  I have created a ticket for this as well: [CXX-2067](https://jira.mongodb.org/browse/CXX-2067).

However, I think that in a real world use case it's likely the user would already have a document containing a value they wanted to encrypt.  I'm imagining that explicit encryption will mostly be used by non-enterprise users, who will write conversion methods to take documents on their way to being inserted, parse a sensitive field out, encrypt that field, then reconstruct the document and insert it.  But who knows, maybe I'm totally wrong about this!  And that use case has its thorny bits, too; we don't have a "replace" facility to drop-in replacement bson values.  In fact, I'm not sure how we would implement that, either...

Reply:
Thanks for creating the tickets! I put some thoughts in CXX-2067.